drivers/flash/soc_flash_nrf.c

```diff
@@ -198,6 +203,8 @@ static int flash_nrf_erase(struct device *dev, off_t addr, size_t size)
 		if (!n_pages) {
 			return 0;
 		}
+
+		addr += DT_FLASH_BASE_ADDRESS;
```

Comment:
@mbolivar, no, the alignment check should be done on the flash area itself, so before adding `DT_FLASH_BASE_ADDRESS`.

Reply:
That doesn't make sense, since it's the actual absolute address at which the erase starts that has to be aligned, right?

I think the answer should be "technically, yes, but as long as `DT_FLASH_BASE_ADDRESS` is page-aligned it doesn't matter, and if it isn't things will break so badly this will be the least of your worries."

I would move the add before the check just for clarity.  (Actually I would have had the address validation function adjust the address, but that's neither here nor there.)

Just an observation: I'm not reviewing this.

Reply:
> That doesn't make sense, since it's the actual absolute address at which the erase starts that has to be aligned, right?
>
Yes, but the absolute address being aligned is the result of two things:
a. The offset being aligned.
b. DT_FLASH_BASE_ADDRESS being aligned.

The flash writing/reading/erasing should only check the first not the second requirement. It would also be wrong to generate an -EINVAL error when DT_FLASH_BASE_ADDRESS has been badly defined.

Reply:
OK, so unpacking this a little, your perspective is that if the argument coming in is correctly aligned then the implementation should neither check for nor diagnose any problem that does not derive from errors in the argument, even though the function invocation will fail to behave as required.

I'd address it a different way (specifically with a build assert that ensures that adding `DT_FLASH_BASE_ADDRESS` maintains alignment, then checking the alignment of the address that's actually going to be used), but what you say makes a certain sense.

It's interesting to see the wide range of perspectives used to guide how Zephyr should be evolved/maintained.  Thanks for explaining.

Reply:
check is done with the assumption that `DT_FLASH_BASE_ADDRESS` is valid as well.
As `page size` is fetched from hal API, it is not possible to make a static assertion for `DT_FLASH_BASE_ADDRESS`.

Reply:
moved the offset adding before the check.

Reply:
This is wrong, if a check is needed it should be added to the init function. Now all writes and reads will perform ok with a bad definition of `DT_FLASH_BASE_ADDRESS`, but only when an erase is done an error will occur.

Even more, I think the introduction of `DT_FLASH_BASE_ADDRESS` is wrong, as soon as this is introduced the number of pages, flash_size, ... need to be changed. As it is the only check required is that `DT_FLASH_BASE_ADDRESS == 0`, as this is supposed in the page layout, size checking, ... this check could easily be added as an assert.

Reply:
^^ added check to the driver initialization.
moved the offset adding after the check.
If check is not satisfied, the driver will be not added to zephyr driver resource, so any attempt to find the driver bindings will fail.


The flash Base address is 0x1000000 for networking CPU in nRF5340.

Reply:
> ^^ added check to the driver initialization.
> moved the offset adding after the check.
> If check is not satisfied, the driver will be not added to zephyr driver resource, so any attempt to find the driver bindings will fail.
> 
> The flash Base address is 0x1000000 for networking CPU in nRF5340.

And what about the size and the page layout ?

Reply:
> And what about the size and the page layout ?

They are fetched from nordic hal - we are not going to check the hal as it comes from nrfx releases with its own verification.

Reply:
Ok, I did some reading on the nRF5340. In this device you have two separate flash controllers, one at 0x41080000 (netcpu) and one at 0x39000 (appcpu + netcpu). The present PR is OK for two separate flash controllers as they each get the flash size and page size from a different controller.

Reply:
^^ should I remove check in initialization then?

Reply:
^^ removed

Reply:
@pabigot can you precise your requirement?

Reply:
IMO, yes.

BTW, have you tried doing a page erase when selecting the netcpu? I think it should fail with an assert.

Reply:
> I think the answer should be "technically, yes, but as long as `DT_FLASH_BASE_ADDRESS` is page-aligned it doesn't matter, and if it isn't things will break so badly this will be the least of your worries."

This seems reasonable to me and the conversation has otherwise gotten into the weeds IMO.

Reply:
> BTW, have you tried doing a page erase when selecting the netcpu? I think it should fail with an assert.

@Laczen Yes, I did and it was working.